MAUI Android build crashes when app is reopened from background. It throws the exception: 'Window was already created.'

### Description

Exception: Window was already created.
Source: Microsoft.Maui.Controls'

StackTrace	"   at Microsoft.Maui.Controls.Window.Microsoft.Maui.IWindow.Created() in D:\\a\\_work\\1\\s\\src\\Controls\\src\\Core\\Window\\Window.cs:line 491\n   at Microsoft.Maui.LifecycleEvents.AppHostBuilderExtensions.<>c.<OnConfigureLifeCycle>b__2_0(Activity activity, Bundle bun…"	string

### Steps to Reproduce

Create a Shell app with Flyout and bottom Tabs.
Click the home button on device to move the app to the background.
Now open the app again from background, it will throw the exception: **Window was already created.**

No app can be released with this kind of bug where app cannot be restarted from background which is very basic thing.

### Link to public reproduction project repository

_No response_

### Version with bug

8.0.0-rc.1.9171

### Is this a regression from previous behavior?

Not sure, did not test other versions

### Last version that worked well

Unknown/Other

### Affected platforms

Android

### Affected platform versions

_No response_

### Did you find any workaround?

No, couldn't find any workaround yet. It is quite frustrating and don't know how to fix it.

### Relevant log output

_No response_

I have similar issue at android - not able to tell the exact source of the exception is as discussed here. However, to me it happens when window is requested from foreground service via notification. I could simulate that when open the window and then try to open it again from the notification.

[mono-rt] [ERROR] FATAL UNHANDLED EXCEPTION: System.InvalidOperationException: Window was already created
[mono-rt]    at Microsoft.Maui.Controls.Window.Microsoft.Maui.IWindow.Created() in D:\a\_work\1\s\src\Controls\src\Core\Window\Window.cs:line 491
[mono-rt]    at Microsoft.Maui.LifecycleEvents.AppHostBuilderExtensions.<>c.<OnConfigureLifeCycle>b__2_0(Activity activity, Bundle bundle) in D:\a\_work\1\s\src\Core\src\Hosting\LifecycleEvents\AppHostBuilderExtensions.Android.cs:line 23
[mono-rt]    at Microsoft.Maui.MauiAppCompatActivity.<>c__DisplayClass8_0.<OnPostCreate>b__0(OnPostCreate del) in D:\a\_work\1\s\src\Core\src\Platform\Android\MauiAppCompatActivity.Lifecycle.cs:line 61
[mono-rt]    at Microsoft.Maui.LifecycleEvents.LifecycleEventServiceExtensions.InvokeLifecycleEvents[OnPostCreate](IServiceProvider services, Action`1 action) in D:\a\_work\1\s\src\Core\src\LifecycleEvents\LifecycleEventServiceExtensions.cs:line 31
[mono-rt]    at Microsoft.Maui.MauiAppCompatActivity.OnPostCreate(Bundle savedInstanceState) in D:\a\_work\1\s\src\Core\src\Platform\Android\MauiAppCompatActivity.Lifecycle.cs:line 61
[mono-rt]    at Android.App.Activity.n_OnPostCreate_Landroid_os_Bundle_(IntPtr jnienv, IntPtr native__this, IntPtr native_savedInstanceState) in /Users/runner/work/1/s/xamarin-android/src/Mono.Android/obj/Release/net8.0/android-34/mcw/Android.App.Activity.cs:line 4488
[mono-rt]    at Android.Runtime.JNINativeWrapper.Wrap_JniMarshal_PPL_V(_JniMarshal_PPL_V callback, IntPtr jnienv, IntPtr klazz, IntPtr p0) in /Users/runner/work/1/s/xamarin-android/src/Mono.Android/Android.Runtime/JNINativeWrapper.g.cs:line 125

At android, I am a bit suffering with events risen quickly one after another when clicked at control. Instead of doubleclik it is recognized as two clicks. Might have some connection to occasional crash when trying to open already opened window?

--- a/src/Core/src/Platform/Android/ApplicationExtensions.cs
+++ b/src/Core/src/Platform/Android/ApplicationExtensions.cs
@@ -53,7 +53,7 @@ namespace Microsoft.Maui.Platform
 					$"This window is already associated with an active Activity ({oldActivity.GetType()}). " + 
 					$"Please override CreateWindow on {application.GetType()} "  + 
 					$"to add support for multiple activities https://learn.microsoft.com/en-us/dotnet/maui/fundamentals/windows?view=net-maui-8.0#create-a-window "  + 
-					$"or set the LaunchMode to SingleTask on {activity.GetType()}.");
+					$"or set the LaunchMode to SingleTop on {activity.GetType()}.");
 			}
 
 			activity.SetWindowHandler(window, mauiContext);

--- a/src/Core/src/Platform/Android/ApplicationExtensions.cs
+++ b/src/Core/src/Platform/Android/ApplicationExtensions.cs
@@ -45,6 +45,17 @@ namespace Microsoft.Maui.Platform
 
 			var window = application.CreateWindow(activationState);
 
+			if (window.Handler?.PlatformView is Activity oldActivity && 
+				oldActivity != activity &&
+				!oldActivity.IsDestroyed)
+			{
+				throw new InvalidOperationException(
+					$"This window is already associated with an active Activity ({oldActivity.GetType()}). " + 
+					$"Please override CreateWindow on {application.GetType()} "  + 
+					$"to add support for multiple activities https://learn.microsoft.com/en-us/dotnet/maui/fundamentals/windows?view=net-maui-8.0#create-a-window "  + 
+					$"or set the LaunchMode to SingleTask on {activity.GetType()}.");
+			}
+
 			activity.SetWindowHandler(window, mauiContext);
 		}
 

--- a/src/Core/src/Platform/Android/ApplicationExtensions.cs
+++ b/src/Core/src/Platform/Android/ApplicationExtensions.cs
@@ -52,7 +52,7 @@ namespace Microsoft.Maui.Platform
 				throw new InvalidOperationException(
 					$"This window is already associated with an active Activity ({oldActivity.GetType()}). " + 
 					$"Please override CreateWindow on {application.GetType()} "  + 
-					$"to add support for multiple activities https://learn.microsoft.com/en-us/dotnet/maui/fundamentals/windows?view=net-maui-8.0#create-a-window "  + 
+					$"to add support for multiple activities https://aka.ms/maui-docs-create-window"  + 
 					$"or set the LaunchMode to SingleTop on {activity.GetType()}.");
 			}
 

--- a/src/Core/tests/DeviceTests.Shared/Stubs/CoreApplicationStub.cs
+++ b/src/Core/tests/DeviceTests.Shared/Stubs/CoreApplicationStub.cs
@@ -5,6 +5,8 @@ namespace Microsoft.Maui.DeviceTests.Stubs
 {
 	public class CoreApplicationStub : IApplication
 	{
+		IWindow _singleWindow;
+
 		readonly List<IWindow> _windows = new List<IWindow>();
 
 		public IElementHandler Handler { get; set; }
@@ -17,9 +19,11 @@ namespace Microsoft.Maui.DeviceTests.Stubs
 
 		public IWindow CreateWindow(IActivationState state)
 		{
+			if (_singleWindow is not null)
+				return _singleWindow;
+				
 			_windows.Add(new WindowStub());
-
-			return _windows[0];
+			return _windows.Last();
 		}
 
 		public void OpenWindow(IWindow window)
@@ -33,5 +37,10 @@ namespace Microsoft.Maui.DeviceTests.Stubs
 		}
 
 		public void ThemeChanged() { }
+
+		public void SetSingleWindow(IWindow window)
+		{
+			_singleWindow = window;
+		}
 	}
 }

--- a/src/Core/tests/DeviceTests/Handlers/Window/WindowHandlerTests.Android.cs
+++ b/src/Core/tests/DeviceTests/Handlers/Window/WindowHandlerTests.Android.cs
@@ -1,12 +1,41 @@
 ﻿using System.Threading.Tasks;
 using AndroidX.AppCompat.App;
 using Microsoft.Maui.Controls;
+using Microsoft.Extensions.DependencyInjection;
 using Xunit;
+using Microsoft.Maui.DeviceTests.Stubs;
+using Android.App;
+using System;
 
 namespace Microsoft.Maui.DeviceTests
 {
 	public partial class WindowHandlerTests : CoreHandlerTestBase
 	{
+
+		[Fact]
+		public async Task UsingTheSameWindowThrowsInvalidOperationException()
+		{
+			await InvokeOnMainThreadAsync(() =>
+			{
+				var app = (CoreApplicationStub)MauiContext.Services.GetRequiredService<IApplication>();
+
+				var activity1 = new MauiAppCompatActivity();
+				var activity2 = new MauiAppCompatActivity();
+				
+				activity1.CreatePlatformWindow(app, null);
+
+				var window = app.Windows[0];
+				app.SetSingleWindow(window);
+
+				Assert.Throws<InvalidOperationException>(() =>
+				{
+					activity2.CreatePlatformWindow(app, null);
+				});
+
+			});
+		}
+
+
 		[Fact]
 		public async Task TitleSetsOnWindow()
 		{

--- a/src/Core/tests/DeviceTests/Handlers/Window/WindowHandlerTests.Android.cs
+++ b/src/Core/tests/DeviceTests/Handlers/Window/WindowHandlerTests.Android.cs
@@ -6,6 +6,7 @@ using Xunit;
 using Microsoft.Maui.DeviceTests.Stubs;
 using Android.App;
 using System;
+using Microsoft.Maui.Hosting;
 
 namespace Microsoft.Maui.DeviceTests
 {
@@ -15,9 +16,20 @@ namespace Microsoft.Maui.DeviceTests
 		[Fact]
 		public async Task UsingTheSameWindowThrowsInvalidOperationException()
 		{
+			EnsureHandlerCreated(builder =>
+			{
+				builder.ConfigureMauiHandlers(handlers =>
+				{
+					handlers.AddHandler<WindowStub, WindowHandlerProxyStub>();
+				});
+			});
+
 			await InvokeOnMainThreadAsync(() =>
 			{
 				var app = (CoreApplicationStub)MauiContext.Services.GetRequiredService<IApplication>();
+				var handler = new ApplicationHandler();
+				app.Handler = handler;
+				handler.SetMauiContext(MauiContext);
 
 				var activity1 = new MauiAppCompatActivity();
 				var activity2 = new MauiAppCompatActivity();

--- a/src/Core/tests/DeviceTests/Stubs/WindowHandlerProxyStub.cs
+++ b/src/Core/tests/DeviceTests/Stubs/WindowHandlerProxyStub.cs
@@ -14,6 +14,11 @@ namespace Microsoft.Maui.DeviceTests.Stubs
 {
 	public class WindowHandlerProxyStub : ElementHandler<IWindow, PlatformView>, IWindowHandler
 	{
+		public WindowHandlerProxyStub() : this(new PropertyMapper<IWindow, IWindowHandler>(), new CommandMapper<IWindow, IWindowHandler>())
+		{
+
+		}
+
 		public WindowHandlerProxyStub(IPropertyMapper<IWindow, IWindowHandler> mapper = null, CommandMapper<IWindow, IWindowHandler> commandMapper = null)
 			: base(mapper ?? new PropertyMapper<IWindow, IWindowHandler>(), commandMapper ?? new CommandMapper<IWindow, IWindowHandler>())
 		{

--- a/src/Templates/src/templates/maui-mobile/Platforms/Android/MainActivity.cs
+++ b/src/Templates/src/templates/maui-mobile/Platforms/Android/MainActivity.cs
@@ -4,7 +4,7 @@ using Android.OS;
 
 namespace MauiApp._1;
 
-[Activity(Theme = "@style/Maui.SplashTheme", MainLauncher = true, LaunchMode = LaunchMode.SingleTask, ConfigurationChanges = ConfigChanges.ScreenSize | ConfigChanges.Orientation | ConfigChanges.UiMode | ConfigChanges.ScreenLayout | ConfigChanges.SmallestScreenSize | ConfigChanges.Density)]
+[Activity(Theme = "@style/Maui.SplashTheme", MainLauncher = true, LaunchMode = LaunchMode.SingleTop, ConfigurationChanges = ConfigChanges.ScreenSize | ConfigChanges.Orientation | ConfigChanges.UiMode | ConfigChanges.ScreenLayout | ConfigChanges.SmallestScreenSize | ConfigChanges.Density)]
 public class MainActivity : MauiAppCompatActivity
 {
 }

--- a/src/Templates/src/templates/maui-mobile/Platforms/Android/MainActivity.cs
+++ b/src/Templates/src/templates/maui-mobile/Platforms/Android/MainActivity.cs
@@ -4,7 +4,7 @@ using Android.OS;
 
 namespace MauiApp._1;
 
-[Activity(Theme = "@style/Maui.SplashTheme", MainLauncher = true, ConfigurationChanges = ConfigChanges.ScreenSize | ConfigChanges.Orientation | ConfigChanges.UiMode | ConfigChanges.ScreenLayout | ConfigChanges.SmallestScreenSize | ConfigChanges.Density)]
+[Activity(Theme = "@style/Maui.SplashTheme", MainLauncher = true, LaunchMode = LaunchMode.SingleTask, ConfigurationChanges = ConfigChanges.ScreenSize | ConfigChanges.Orientation | ConfigChanges.UiMode | ConfigChanges.ScreenLayout | ConfigChanges.SmallestScreenSize | ConfigChanges.Density)]
 public class MainActivity : MauiAppCompatActivity
 {
 }

--- a/src/Templates/src/templates/maui-multiproject/MauiApp.1.Droid/MainActivity.cs
+++ b/src/Templates/src/templates/maui-multiproject/MauiApp.1.Droid/MainActivity.cs
@@ -4,7 +4,7 @@ using Android.OS;
 
 namespace MauiApp._1.Droid;
 
-[Activity(Theme = "@style/Maui.SplashTheme", MainLauncher = true, LaunchMode = LaunchMode.SingleTask, ConfigurationChanges = ConfigChanges.ScreenSize | ConfigChanges.Orientation | ConfigChanges.UiMode | ConfigChanges.ScreenLayout | ConfigChanges.SmallestScreenSize | ConfigChanges.Density)]
+[Activity(Theme = "@style/Maui.SplashTheme", MainLauncher = true, LaunchMode = LaunchMode.SingleTop, ConfigurationChanges = ConfigChanges.ScreenSize | ConfigChanges.Orientation | ConfigChanges.UiMode | ConfigChanges.ScreenLayout | ConfigChanges.SmallestScreenSize | ConfigChanges.Density)]
 public class MainActivity : MauiAppCompatActivity
 {
 }

--- a/src/Templates/src/templates/maui-multiproject/MauiApp.1.Droid/MainActivity.cs
+++ b/src/Templates/src/templates/maui-multiproject/MauiApp.1.Droid/MainActivity.cs
@@ -4,7 +4,7 @@ using Android.OS;
 
 namespace MauiApp._1.Droid;
 
-[Activity(Theme = "@style/Maui.SplashTheme", MainLauncher = true, ConfigurationChanges = ConfigChanges.ScreenSize | ConfigChanges.Orientation | ConfigChanges.UiMode | ConfigChanges.ScreenLayout | ConfigChanges.SmallestScreenSize | ConfigChanges.Density)]
+[Activity(Theme = "@style/Maui.SplashTheme", MainLauncher = true, LaunchMode = LaunchMode.SingleTask, ConfigurationChanges = ConfigChanges.ScreenSize | ConfigChanges.Orientation | ConfigChanges.UiMode | ConfigChanges.ScreenLayout | ConfigChanges.SmallestScreenSize | ConfigChanges.Density)]
 public class MainActivity : MauiAppCompatActivity
 {
 }

